# FELIPE GRUMSER FRANCO VELHO
## MPV DA DISCIPLINA ENGENHARIA DE DADOS

O presente trabalho tem como objetivo criar um modelo de dados percorrendo pipelines de ETL utilizando a linguagem SQL. Ele está estruturado nas camadas bronze, prata e ouro. A carga de dados é feita na camada bronze, a análise da qualidade de manipulação dos dados ocorrerá na camada prata e as consultas na camada Gold.

A base de dados escolhida foi a publicamente disponibilizada pela IMDb (Internet Movie Database), que contém informações sobre filmes, séries e documentários. Com base nos dados disponibilizados, objetivou-se tratá-los a fim de obter, na camada Gold, mecanismos para responder perguntas sobre as relações de ano e gênero. As perguntas que se deseja responder são:
- Qual o ranking dos gêneros dos filmes lançados no ano X com base na média das suas notas?
- Quais as Top 10 melhores notas e seus respectivos ano e gênero?
- Dado um gênero X, qual o Top 50 melhores notas e seus respectivos anos?
- Quais são os anos com o maior número de votos, independente de gênero?


O trecho de código a seguir realiza a importação das bibliotecas necessárias para a manipulação e obtenção de dados. A biblioteca pandas, importada como pd, é utilizada para análise e manipulação de dados em formato tabular. A biblioteca io permite o manuseio de fluxos de dados em memória. Já a biblioteca requests possibilita a realização de requisições HTTP, sendo útil para acessar dados hospedados em fontes externas, como os arquivos disponibilizados pelo IMBd, via url. Para esse projeto, foi importada a tabela contendo o título original e primário das obras, seu tipo, se é conteúdo adulto, ano de início, tempo de exibição e gênero. A outra tabela importada contém a média de avaliações de cada título e o número de votos. Além disso, todos os filmes possuem um identificador "tconst". 

In [0]:
import pandas as pd
import io
import requests

In [0]:
url = "https://datasets.imdbws.com/title.basics.tsv.gz"
title_basics_df = pd.read_csv(url, compression='gzip', sep='\t')

title_basics_df.head()

<command-1370257541152233>:2: DtypeWarning: Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.
  title_basics_df = pd.read_csv(url, compression='gzip', sep='\t')


tconst 
 titleType 
 primaryTitle 
 originalTitle 
 isAdult 
 startYear 
 endYear 
 runtimeMinutes 
 genres 
 
 
 
 
 0 
 tt0000001 
 short 
 Carmencita 
 Carmencita 
 0 
 1894 
 \N 
 1 
 Documentary,Short 
 
 
 1 
 tt0000002 
 short 
 Le clown et ses chiens 
 Le clown et ses chiens 
 0 
 1892 
 \N 
 5 
 Animation,Short 
 
 
 2 
 tt0000003 
 short 
 Poor Pierrot 
 Pauvre Pierrot 
 0 
 1892 
 \N 
 5 
 Animation,Comedy,Romance 
 
 
 3 
 tt0000004 
 short 
 Un bon bock 
 Un bon bock 
 0 
 1892 
 \N 
 12 
 Animation,Short 
 
 
 4 
 tt0000005 
 short 
 Blacksmith Scene 
 Blacksmith Scene 
 0 
 1893 
 \N 
 1 
 Short

In [0]:
url = "https://datasets.imdbws.com/title.ratings.tsv.gz"
title_ratings_df = pd.read_csv(url, compression='gzip', sep='\t')

title_ratings_df.head()

tconst 
 averageRating 
 numVotes 
 
 
 
 
 0 
 tt0000001 
 5.7 
 2146 
 
 
 1 
 tt0000002 
 5.5 
 291 
 
 
 2 
 tt0000003 
 6.5 
 2180 
 
 
 3 
 tt0000004 
 5.3 
 187 
 
 
 4 
 tt0000005 
 6.2 
 2924

Como mencionado anteriormente, o trabalho está dividido em categorias, começando pela camada bronze. Primeiramente, realizou-se a exclusão da camada para garantir que não houvesse nenhuma tabela previamente criada com o mesmo nome e a recriou.

In [0]:
%sql
DROP DATABASE bronze;

---------------------------------------------------------------------------
AnalysisException                         Traceback (most recent call last)
File <command-1370257541152238>:7
      5     display(df)
      6     return df
----> 7   _sqldf = ____databricks_percent_sql()
      8 finally:
      9   del ____databricks_percent_sql

File <command-1370257541152238>:4, in ____databricks_percent_sql()
      2 def ____databricks_percent_sql():
      3   import base64
----> 4   df = spark.sql(base64.standard_b64decode("RFJPUCBEQVRBQkFTRSBicm9uemU=").decode())
      5   display(df)
      6   return df

File /databricks/spark/python/pyspark/instrumentation_utils.py:48, in _wrap_function.<locals>.wrapper(*args, **kwargs)
     46 start = time.perf_counter()
     47 try:
---> 48     res = func(*args, **kwargs)
     49     logger.log_success(
     50         module_name, class_name, function_name, time.perf_counter() - start, signature
     51     )
     52     return res

File /databricks/sp

In [0]:
%sql
CREATE DATABASE bronze;

A coluna isAdult apresentou um erro ao ser criada no dataframe Spark. Como os seus dados não seriam úteis para resoponder as perguntas, optou-se por removê-la para facilitar o processo.

In [0]:
title_basics_df = title_basics_df.drop(columns=["isAdult"])

Os blocos de código a seguir estão criando as tabelas "title_basics" e "title_ratings" na camada bronze do banco de dados. O Apache Spark permite manipulação de dados de forma eficiente, sendo essencial para a criação do dataframe. A primeira linha é responsável por, a partir do dataframe gerado pela importação do Pandas, criar um dataframe no Spark. Em seguida, o data frame do Spark é salvo na camada bronze como tabela, escrevendo por cima de possível arquivo de mesmo nome.

In [0]:
title_basics_spark_df = spark.createDataFrame(title_basics_df)
title_basics_spark_df.write.mode("overwrite").saveAsTable("bronze.title_basics")

In [0]:
title_ratings_spark_df = spark.createDataFrame(title_ratings_df)
title_ratings_spark_df.write.mode("overwrite").saveAsTable("bronze.title_ratings")

In [0]:
%sql
SELECT * FROM bronze.title_basics LIMIT 20

tconst titleType primaryTitle originalTitle startYear endYear runtimeMinutes genres tt0000001 short Carmencita Carmencita 1894 \N 1 Documentary,Short tt0000002 short Le clown et ses chiens Le clown et ses chiens 1892 \N 5 Animation,Short tt0000003 short Poor Pierrot Pauvre Pierrot 1892 \N 5 Animation,Comedy,Romance tt0000004 short Un bon bock Un bon bock 1892 \N 12 Animation,Short tt0000005 short Blacksmith Scene Blacksmith Scene 1893 \N 1 Short tt0000006 short Chinese Opium Den Chinese Opium Den 1894 \N 1 Short tt0000007 short Corbett and Courtney Before the Kinetograph Corbett and Courtney Before the Kinetograph 1894 \N 1 Short,Sport tt0000008 short Edison Kinetoscopic Record of a Sneeze Edison Kinetoscopic Record of a Sneeze 1894 \N 1 Documentary,Short tt0000009 movie Miss Jerry Miss Jerry 1894 \N 45 Romance tt0000010 short Leaving the Factory La sortie de l'usine Lumière à Lyon 1895 \N 1 Documentary,Short tt0000011 short Akrobatisches Potpourri Akrobatisches Potpourri 1895 \N 1 Documentary,Short tt0000012 short The Arrival of a Train L'arrivée d'un train à La Ciotat 1896 \N 1 Documentary,Short tt0000013 short The Photographical Congress Arrives in Lyon Le débarquement du congrès de photographie à Lyon 1895 \N 1 Documentary,Short tt0000014 short The Waterer Watered L'arroseur arrosé 1895 \N 1 Comedy,Short tt0000015 short Around a Cabin Autour d'une cabine 1894 \N 2 Animation,Comedy,Short tt0000016 short Boat Leaving the Port Barque sortant du port 1895 \N 1 Documentary,Short tt0000017 short Italienischer Bauerntanz Italienischer Bauerntanz 1895 \N 1 Documentary,Short tt0000018 short Das boxende Känguruh Das boxende Känguruh 1895 \N 1 Short,Sport tt0000019 short The Clown Barber The Clown Barber 1898 \N \N Comedy,Short tt0000020 short The Derby 1895 The Derby 1895 1895 \N 1 Documentary,Short,Sport

Aqui, visualizou-se os tipos de títulos disponíveis com o comando SELECT DISTINCT, que mostra os valores únicos da coluna selecionada. Como o presente trabalho vai analisar apenas filmes, apenas os tipos Movie, tvShort, short e tvMovie serão selecionados posteriormente.

In [0]:
%sql
SELECT DISTINCT titleType FROM bronze.title_basics;

titleType tvSeries tvMiniSeries tvMovie tvEpisode movie tvSpecial video videoGame tvShort short tvPilot

In [0]:
%sql
SELECT * FROM bronze.title_ratings LIMIT 20

tconst averageRating numVotes tt0000001 5.7 2146 tt0000002 5.5 291 tt0000003 6.5 2180 tt0000004 5.3 187 tt0000005 6.2 2924 tt0000006 5.0 210 tt0000007 5.3 906 tt0000008 5.4 2292 tt0000009 5.4 223 tt0000010 6.8 7922 tt0000011 5.2 418 tt0000012 7.4 13447 tt0000013 5.7 2064 tt0000014 7.1 6138 tt0000015 6.1 1271 tt0000016 5.9 1656 tt0000017 4.6 379 tt0000018 5.2 662 tt0000019 5.2 35 tt0000020 4.7 410

Uma vez esgotado as ações a serem realizadas na camada bronze (importação de dados) cria-se de maneira similar a camada prata para a análise dos dados, verificando a sua consistência.

In [0]:
%sql
DROP DATABASE silver CASCADE

---------------------------------------------------------------------------
AnalysisException                         Traceback (most recent call last)
File <command-1370257541152250>:7
      5     display(df)
      6     return df
----> 7   _sqldf = ____databricks_percent_sql()
      8 finally:
      9   del ____databricks_percent_sql

File <command-1370257541152250>:4, in ____databricks_percent_sql()
      2 def ____databricks_percent_sql():
      3   import base64
----> 4   df = spark.sql(base64.standard_b64decode("RFJPUCBEQVRBQkFTRSBzaWx2ZXIgQ0FTQ0FERQ==").decode())
      5   display(df)
      6   return df

File /databricks/spark/python/pyspark/instrumentation_utils.py:48, in _wrap_function.<locals>.wrapper(*args, **kwargs)
     46 start = time.perf_counter()
     47 try:
---> 48     res = func(*args, **kwargs)
     49     logger.log_success(
     50         module_name, class_name, function_name, time.perf_counter() - start, signature
     51     )
     52     return res

File /d

In [0]:
%sql
CREATE DATABASE silver

In [0]:
%sql
CREATE TABLE silver.title_ratings AS
SELECT * FROM bronze.title_ratings;

num_affected_rows num_inserted_rows

O primeiro passo foi usar a explosão para separar os gêneros que originalmente estavam separados por vírgula na coluna "genre". Dessa forma, as obras que possuem mais de um gênero ocuparão mais de uma linha. Por esse motivo, o identificador único "tcosnt" perdeu a sua unicidade. Além disso, foram descatadas algumas colunas que não seriam relevantes para o projeto.

In [0]:
from pyspark.sql.functions import explode, split

title_genres_exploded_df = title_basics_spark_df.select(
    "tconst",
    "primaryTitle",
    "titleType",
    "startYear",
    explode(split("genres", ",")).alias("genre")
)

title_genres_exploded_df.show(10)


+---------+--------------------+---------+---------+-----------+
|   tconst|        primaryTitle|titleType|startYear|      genre|
+---------+--------------------+---------+---------+-----------+
|tt0000001|          Carmencita|    short|     1894|Documentary|
|tt0000001|          Carmencita|    short|     1894|      Short|
|tt0000002|Le clown et ses c...|    short|     1892|  Animation|
|tt0000002|Le clown et ses c...|    short|     1892|      Short|
|tt0000003|        Poor Pierrot|    short|     1892|  Animation|
|tt0000003|        Poor Pierrot|    short|     1892|     Comedy|
|tt0000003|        Poor Pierrot|    short|     1892|    Romance|
|tt0000004|         Un bon bock|    short|     1892|  Animation|
|tt0000004|         Un bon bock|    short|     1892|      Short|
|tt0000005|    Blacksmith Scene|    short|     1893|      Short|
+---------+--------------------+---------+---------+-----------+
only showing top 10 rows



In [0]:
title_genres_exploded_df.write.mode("overwrite").saveAsTable("silver.title_basics")

In [0]:
%sql
SELECT * FROM silver.title_basics LIMIT 100

tconst primaryTitle titleType startYear genre tt0000001 Carmencita short 1894 Documentary tt0000001 Carmencita short 1894 Short tt0000002 Le clown et ses chiens short 1892 Animation tt0000002 Le clown et ses chiens short 1892 Short tt0000003 Poor Pierrot short 1892 Animation tt0000003 Poor Pierrot short 1892 Comedy tt0000003 Poor Pierrot short 1892 Romance tt0000004 Un bon bock short 1892 Animation tt0000004 Un bon bock short 1892 Short tt0000005 Blacksmith Scene short 1893 Short tt0000006 Chinese Opium Den short 1894 Short tt0000007 Corbett and Courtney Before the Kinetograph short 1894 Short tt0000007 Corbett and Courtney Before the Kinetograph short 1894 Sport tt0000008 Edison Kinetoscopic Record of a Sneeze short 1894 Documentary tt0000008 Edison Kinetoscopic Record of a Sneeze short 1894 Short tt0000009 Miss Jerry movie 1894 Romance tt0000010 Leaving the Factory short 1895 Documentary tt0000010 Leaving the Factory short 1895 Short tt0000011 Akrobatisches Potpourri short 1895 Documentary tt0000011 Akrobatisches Potpourri short 1895 Short tt0000012 The Arrival of a Train short 1896 Documentary tt0000012 The Arrival of a Train short 1896 Short tt0000013 The Photographical Congress Arrives in Lyon short 1895 Documentary tt0000013 The Photographical Congress Arrives in Lyon short 1895 Short tt0000014 The Waterer Watered short 1895 Comedy tt0000014 The Waterer Watered short 1895 Short tt0000015 Around a Cabin short 1894 Animation tt0000015 Around a Cabin short 1894 Comedy tt0000015 Around a Cabin short 1894 Short tt0000016 Boat Leaving the Port short 1895 Documentary tt0000016 Boat Leaving the Port short 1895 Short tt0000017 Italienischer Bauerntanz short 1895 Documentary tt0000017 Italienischer Bauerntanz short 1895 Short tt0000018 Das boxende Känguruh short 1895 Short tt0000018 Das boxende Känguruh short 1895 Sport tt0000019 The Clown Barber short 1898 Comedy tt0000019 The Clown Barber short 1898 Short tt0000020 The Derby 1895 short 1895 Documentary tt0000020 The Derby 1895 short 1895 Short tt0000020 The Derby 1895 short 1895 Sport tt0000022 Blacksmith Scene short 1895 Documentary tt0000022 Blacksmith Scene short 1895 Short tt0000023 The Sea short 1895 Documentary tt0000023 The Sea short 1895 Short tt0000024 Opening of the Kiel Canal short 1895 News tt0000024 Opening of the Kiel Canal short 1895 Short tt0000025 The Oxford and Cambridge University Boat Race short 1896 News tt0000025 The Oxford and Cambridge University Boat Race short 1896 Short tt0000025 The Oxford and Cambridge University Boat Race short 1896 Sport tt0000026 The Messers. Lumière at Cards short 1896 Documentary tt0000026 The Messers. Lumière at Cards short 1896 Short tt0000027 Cordeliers' Square in Lyon short 1895 Documentary tt0000027 Cordeliers' Square in Lyon short 1895 Short tt0000028 Fishing for Goldfish short 1895 Documentary tt0000028 Fishing for Goldfish short 1895 Short tt0000029 Baby's Meal short 1895 Documentary tt0000029 Baby's Meal short 1895 Short tt0000030 Rough Sea at Dover short 1895 Documentary tt0000030 Rough Sea at Dover short 1895 Short tt0000031 Jumping the Blanket short 1895 Documentary tt0000031 Jumping the Blanket short 1895 Short tt0000032 Die Serpentintänzerin short 1895 Short tt0000033 Horse Trick Riders short 1895 Comedy tt0000033 Horse Trick Riders short 1895 Documentary tt0000033 Horse Trick Riders short 1895 Short tt0000034 Arrivée d'un train gare de Vincennes short 1896 Documentary tt0000034 Arrivée d'un train gare de Vincennes short 1896 Short tt0000035 Watering the Flowers short 1896 Comedy tt0000035 Watering the Flowers short 1896 Short tt0000036 Awakening of Rip short 1896 Drama tt0000036 Awakening of Rip short 1896 Short tt0000037 Sea Bathing short 1896 Short tt0000038 The Ball Game short 1898 Documentary tt0000038 The Ball Game short 1898 Short tt0000038 The Ball Game short 1898 Sport tt0000039 Barnet Horse Fair short 1896 Short tt0000040 Barque sortant du port de Trouville short 1896 Documentary tt0000040 Barque sortant 

A partir desse ponto, as tabelas são submetidas a uma série de consultas, utilizando os comando SQL, para verificar a qualidade dos dados.

In [0]:
%sql
SELECT COUNT(*) 
FROM silver.title_basics 
WHERE tconst IS NULL;

count(1) 0

Não foram encontrados valores nulos em tconst.

In [0]:
%sql
SELECT *
FROM silver.title_basics
WHERE primaryTitle IS NULL OR TRIM(primaryTitle) = '' OR primaryTitle = '\N'


tconst primaryTitle titleType startYear genre tt5337950 N tvEpisode 2012 Comedy tt5773048 null tvEpisode 2015 Talk-Show tt29387614 N short 2011 Drama tt29387614 N short 2011 Music tt29387614 N short 2011 Short tt33094763 null tvEpisode 2019 Drama tt31462159 null tvEpisode 2024 Drama tt4491564 N short 1978 Short tt19316378 N tvEpisode 2010 Music tt15700278 null tvEpisode 2021 Talk-Show tt1321439 N movie 2005 Documentary tt32915221 N tvEpisode 2021 Comedy tt32915221 N tvEpisode 2021 Romance tt5102240 N tvEpisode 1957 Comedy tt5102240 N tvEpisode 1957 Family tt5102240 N tvEpisode 1957 Music tt5105070 N tvEpisode 1959 Comedy tt5105070 N tvEpisode 1959 Family tt5105070 N tvEpisode 1959 Music tt7869618 N short 2018 Fantasy tt7869618 N short 2018 Sci-Fi tt7869618 N short 2018 Short tt1403289 N tvEpisode 2006 Music tt8312748 N short 2017 Short tt3874996 N tvEpisode 1988 Comedy tt3874996 N tvEpisode 1988 Family tt3874996 N tvEpisode 1988 Music tt2305914 null tvEpisode \N Comedy tt2305914 null tvEpisode \N Talk-Show tt2305918 null tvEpisode \N Comedy tt2305918 null tvEpisode \N Talk-Show tt8541336 null tvEpisode 2018 Reality-TV tt8541336 null tvEpisode 2018 Romance tt2381676 N tvEpisode 1977 Family tt14777874 N tvEpisode 2021 Family tt14510930 null tvEpisode \N \N tt8473688 null tvEpisode 1987 Drama tt21883066 null tvEpisode 2022 News tt21883066 null tvEpisode 2022 Talk-Show tt2067043 null tvEpisode 1965 Music tt9356778 N movie \N Sci-Fi tt9469474 N videoGame 2004 Action tt9469474 N videoGame 2004 Adventure tt1971246 null tvEpisode 2011 Biography

Foram identificadas 44 linhas com inconsistências na coluna primaryTitle e optou-se por alterar esses valores para "Unknown".

In [0]:
%sql
UPDATE silver.title_basics
SET primaryTitle = 'Unknown'
WHERE (primaryTitle IS NULL OR primaryTitle = '\N' OR TRIM(primaryTitle) = '')

num_affected_rows 44

In [0]:
%sql
SELECT COUNT(*) 
FROM silver.title_basics
WHERE titleType IS NULL OR TRIM(titleType) = '' OR primaryTitle = '\\N';

count(1) 0

Não foram encontrados valores nulos em titleType.

In [0]:
%sql
SELECT COUNT(*)
FROM silver.title_basics 
WHERE startYear IS NULL OR startYear = '\\N';

count(1) 2097887

In [0]:
%sql
DELETE FROM silver.title_basics 
WHERE startYear IS NULL OR startYear = '\\N';

num_affected_rows 2097887

Optou-se por deletar todos os filmes com valores de startYear nulos já que o dado é essencial para a análise.

In [0]:
%sql
SELECT COUNT(*)
FROM silver.title_basics 
WHERE startYear = ' ';

count(1) 0

In [0]:
%sql
ALTER TABLE silver.title_basics 
ADD COLUMN startYear_int INT;

UPDATE silver.title_basics 
SET startYear_int = CAST(startYear AS INT);

num_affected_rows 16369220

Os valores de startYear estavam armazenados como string, sendo necessária a sua conversão para INT a fim de realizar as análises.

In [0]:
%sql
SELECT MIN(startYear_int), MAX(startYear_int) 
FROM silver.title_basics

min(startYear_int) max(startYear_int) 1874 2031

In [0]:
%sql
SELECT COUNT(*)
FROM silver.title_basics
WHERE startYear_int > 2024

count(1) 147621

Como o IMDB cataloga filmes que ainda não foram lançados, optou-se por excluir qualquer filme com startYear de 2025 em diante.

In [0]:
%sql
DELETE FROM silver.title_basics
WHERE startYear_int > 2024;

num_affected_rows 147621

In [0]:
%sql
SELECT COUNT(*) 
FROM silver.title_basics 
WHERE genre IS NULL OR genre = '\\N' OR TRIM(genre) = '';


count(1) 421303

In [0]:
%sql
DELETE FROM silver.title_basics 
WHERE genre IS NULL OR genre = '\\N' OR TRIM(genre) = '';

num_affected_rows 421303

Como gênero é um dado essencial para a análise, as tuplas sem esse dado foram deletadas.

In [0]:
%sql
SELECT genre, COUNT(*) 
FROM silver.title_basics 
GROUP BY genre 
ORDER BY COUNT(*) DESC;


genre count(1) Drama 2692487 Comedy 1965639 Talk-Show 1228376 Short 1185111 Documentary 1041027 News 980943 Romance 937987 Family 715930 Reality-TV 575510 Animation 496381 Crime 427900 Action 426736 Game-Show 396793 Adventure 396040 Music 380588 Adult 366579 Sport 277870 Fantasy 219520 Mystery 217669 Horror 198628 Thriller 157599 History 156955 Biography 111168 Sci-Fi 105735 Musical 73852 War 36972 Western 29437 Film-Noir 864

Há um número expressivo de filmes em cada um dos gênero

In [0]:
%sql
SHOW TABLES IN silver;

database tableName isTemporary silver title_basics false silver title_ratings false

Terminada as verificações na tabela com os títulos, iniciou-se as na tabela com as avaliações

In [0]:
%sql
SELECT *
FROM silver.title_ratings LIMIT 50;

tconst averageRating numVotes tt0000001 5.7 2146 tt0000002 5.5 291 tt0000003 6.5 2180 tt0000004 5.3 187 tt0000005 6.2 2924 tt0000006 5.0 210 tt0000007 5.3 906 tt0000008 5.4 2292 tt0000009 5.4 223 tt0000010 6.8 7922 tt0000011 5.2 418 tt0000012 7.4 13447 tt0000013 5.7 2064 tt0000014 7.1 6138 tt0000015 6.1 1271 tt0000016 5.9 1656 tt0000017 4.6 379 tt0000018 5.2 662 tt0000019 5.2 35 tt0000020 4.7 410 tt0000022 5.1 1209 tt0000023 5.7 1584 tt0000024 3.7 158 tt0000025 4.2 51 tt0000026 5.6 1674 tt0000027 5.6 1270 tt0000028 5.1 1197 tt0000029 5.9 3668 tt0000030 5.2 960 tt0000031 5.5 1095 tt0000032 4.9 461 tt0000033 5.5 1132 tt0000034 4.8 253 tt0000035 5.5 88 tt0000036 4.4 654 tt0000037 4.5 80 tt0000038 4.1 232 tt0000039 3.9 36 tt0000040 4.2 82 tt0000041 6.7 2111 tt0000042 4.0 42 tt0000043 3.9 36 tt0000044 4.1 57 tt0000045 4.2 37 tt0000046 4.3 42 tt0000047 3.9 41 tt0000048 4.7 234 tt0000049 4.9 57 tt0000050 4.3 38 tt0000051 3.5 39

In [0]:
%sql
SELECT COUNT(*)
FROM silver.title_ratings
WHERE tconst IS NULL OR TRIM(tconst) = '' OR tconst = '\\N';

count(1) 0

Não há valores nulos em tconst

In [0]:
%sql
SELECT COUNT(*)
FROM silver.title_ratings
WHERE averageRating IS NULL OR TRIM(averageRating) = '';

count(1) 0

Não há valores nulos em avarageRating

In [0]:
%sql
SELECT COUNT(*)
FROM silver.title_ratings
WHERE numVotes IS NULL OR TRIM(numVotes) = '';

count(1) 0

Não há valores nulos em numVotes

In [0]:
%sql
SELECT COUNT(*)
FROM silver.title_ratings
WHERE averageRating < 0 OR averageRating > 10;

count(1) 0

Não há notas fora do intervalo de 0 a 10

In [0]:
%sql
SELECT COUNT(*)
FROM silver.title_ratings
WHERE numVotes = 0;

count(1) 0

Não há filmes com 0 número de votos

In [0]:
%sql
SELECT COUNT(*)
FROM silver.title_ratings r
LEFT JOIN silver.title_basics b ON r.tconst = b.tconst
WHERE b.tconst IS NULL;

count(1) 33294

Verificando a correspondência dos títulos nas duas tabelas com o comando JOIN.

As duas tabelas serão mergidas usando o comando JOIN pelo valor tconst

In [0]:
%sql
CREATE TABLE silver.title_joined AS
SELECT 
    b.tconst,
    b.primaryTitle,
    b.titleType,
    b.startYear,
    b.genre,
    r.averageRating,
    r.numVotes
FROM 
    silver.title_basics b
JOIN 
    silver.title_ratings r
ON 
    b.tconst = r.tconst;


num_affected_rows num_inserted_rows

In [0]:
%sql
SELECT *
FROM silver.title_joined LIMIT 100;

tconst primaryTitle titleType startYear genre averageRating numVotes tt0000008 Edison Kinetoscopic Record of a Sneeze short 1894 Documentary 5.4 2292 tt0000008 Edison Kinetoscopic Record of a Sneeze short 1894 Short 5.4 2292 tt0000015 Around a Cabin short 1894 Animation 6.1 1271 tt0000015 Around a Cabin short 1894 Comedy 6.1 1271 tt0000015 Around a Cabin short 1894 Short 6.1 1271 tt0000017 Italienischer Bauerntanz short 1895 Documentary 4.6 379 tt0000017 Italienischer Bauerntanz short 1895 Short 4.6 379 tt0000019 The Clown Barber short 1898 Comedy 5.2 35 tt0000019 The Clown Barber short 1898 Short 5.2 35 tt0000051 The Bohemian Encampment short 1896 Documentary 3.5 39 tt0000051 The Bohemian Encampment short 1896 Short 3.5 39 tt0000058 Cortège de tzar au Bois de Boulogne short 1896 Documentary 4.4 37 tt0000058 Cortège de tzar au Bois de Boulogne short 1896 Short 4.4 37 tt0000065 Dessinateur: Reine Victoria short 1896 Short 3.5 33 tt0000095 The Mysterious Paper short 1896 Short 4.8 38 tt0000108 Rip Leaving Sleepy Hollow short 1896 Drama 4.4 588 tt0000108 Rip Leaving Sleepy Hollow short 1896 Short 4.4 588 tt0000142 Buffalo Bill and Escort short 1897 News 5.0 34 tt0000142 Buffalo Bill and Escort short 1897 Short 5.0 34 tt0000145 En classe short 1897 Comedy 3.9 16 tt0000145 En classe short 1897 Short 3.9 16 tt0000146 Le cocher de fiacre endormi short 1897 Comedy 4.8 29 tt0000146 Le cocher de fiacre endormi short 1897 Short 4.8 29 tt0000169 Slagsmål i gamla Stockholm short 1897 Short 4.0 63 tt0000172 The X-Ray Fiend short 1897 Comedy 6.0 1144 tt0000172 The X-Ray Fiend short 1897 Horror 6.0 1144 tt0000172 The X-Ray Fiend short 1897 Short 6.0 1144 tt0000174 Výstavní párkar a lepic plakátù short 1898 Comedy 4.5 148 tt0000174 Výstavní párkar a lepic plakátù short 1898 Short 4.5 148 tt0000175 A Sea Cave Near Lisbon short 1896 Documentary 4.9 257 tt0000175 A Sea Cave Near Lisbon short 1896 Short 4.9 257 tt0000186 Le cuirassé Maine short 1898 Short 4.6 27 tt0000186 Le cuirassé Maine short 1898 War 4.6 27 tt0000188 The Deserter short 1898 Drama 4.3 14 tt0000188 The Deserter short 1898 Short 4.3 14 tt0000195 Les farces de Jocko short 1898 Comedy 5.4 12 tt0000195 Les farces de Jocko short 1898 Short 5.4 12 tt0000212 Saida do Paquete Duque de Braganca short 1898 Short 5.4 11 tt0000214 Tearing Down the Spanish Flag short 1898 Short 4.0 124 tt0000214 Tearing Down the Spanish Flag short 1898 War 4.0 124 tt0000242 The Sign of the Cross short 1899 Fantasy 6.4 1452 tt0000242 The Sign of the Cross short 1899 Horror 6.4 1452 tt0000242 The Sign of the Cross short 1899 Short 6.4 1452 tt0000264 Summoning the Spirits short 1899 Fantasy 5.7 524 tt0000264 Summoning the Spirits short 1899 Horror 5.7 524 tt0000264 Summoning the Spirits short 1899 Short 5.7 524 tt0000302 La fée aux choux, ou la naissance des enfants short 1900 Short 5.3 325 tt0000305 L'Habanera short 1900 Short 5.8 12 tt0000337 Spanish Bullfight short 1900 Short 4.9 394 tt0000337 Spanish Bullfight short 1900 Sport 4.9 394 tt0000357 The Haunted Curiosity Shop short 1901 Comedy 5.8 548 tt0000357 The Haunted Curiosity Shop short 1901 Fantasy 5.8 548 tt0000357 The Haunted Curiosity Shop short 1901 Horror 5.8 548 tt0000358 History of a Crime short 1901 Crime 6.1 832 tt0000358 History of a Crime short 1901 Drama 6.1 832 tt0000358 History of a Crime short 1901 Short 6.1 832 tt0000376 A táncz short 1901 Documentary 6.1 26 tt0000376 A táncz short 1901 Short 6.1 26 tt0000379 Visita de la escuadra inglesa a Barcelona short 1901 Documentary 4.3 16 tt0000379 Visita de la escuadra inglesa a Barcelona short 1901 Short 4.3 16 tt0000400 Le lion savant short 1902 Comedy 6.0 20 tt0000400 Le lion savant short 1902 Short 6.0 20 tt0000401 Le marchand de ballons short 1902 Comedy 5.2 10 tt0000401 Le marchand de ballons short 1902 Short 5.2 10 tt0000409 Midwife to the Upper Classes short 1902 Comedy 5.1 304 tt0000409 Midwife to the Upper Classes short 1902 Short 5.1 304 tt0000422 Les aventures d'un voyageur trop press

Considerando que as atividades a serem realizadas na camada prata foram encerradas, a camada ouro será criada a seguir.

In [0]:
%sql
DROP DATABASE gold CASCADE

---------------------------------------------------------------------------
AnalysisException                         Traceback (most recent call last)
File <command-1370257541152297>:7
      5     display(df)
      6     return df
----> 7   _sqldf = ____databricks_percent_sql()
      8 finally:
      9   del ____databricks_percent_sql

File <command-1370257541152297>:4, in ____databricks_percent_sql()
      2 def ____databricks_percent_sql():
      3   import base64
----> 4   df = spark.sql(base64.standard_b64decode("RFJPUCBEQVRBQkFTRSBnb2xkIENBU0NBREU=").decode())
      5   display(df)
      6   return df

File /databricks/spark/python/pyspark/instrumentation_utils.py:48, in _wrap_function.<locals>.wrapper(*args, **kwargs)
     46 start = time.perf_counter()
     47 try:
---> 48     res = func(*args, **kwargs)
     49     logger.log_success(
     50         module_name, class_name, function_name, time.perf_counter() - start, signature
     51     )
     52     return res

File /datab

In [0]:
%sql
CREATE DATABASE gold

O código a seguir cria a tabela final, que será usada para a consulta. Ele filtra os títulos para incluir apenas os tipos movie, tvShort, short e tvMovie, que serão os considerados na análise. O startYear teve seu nome alterado para year e foi calculada a média de avaliações e a soma do número de votos por ano e gênero. Para isso, o código agrupou startYear e genre, juntando os registros em que esse valores são iguais. Por fim, ordena o resultado por ano e gênero.

In [0]:
%sql
CREATE TABLE gold.genre_analysis AS
SELECT
    startYear AS year,
    genre,
    AVG(CAST(averageRating AS DOUBLE)) AS avgRating, 
    SUM(CAST(numVotes AS INT)) AS totalVotes
FROM
    silver.title_joined
WHERE
    titleType IN ('movie', 'tvShort', 'short', 'tvMovie')
GROUP BY
    startYear,
    genre
ORDER BY
    startYear,
    genre;


num_affected_rows num_inserted_rows

In [0]:
%sql
SELECT *
FROM gold.genre_analysis
WHERE year > 1980 LIMIT 500;

year genre avgRating totalVotes 1981 Action 5.807232704402517 1943670 1981 Adult 5.490948275862069 17224 1981 Adventure 6.039784946236559 2135619 1981 Animation 6.554587155963303 200103 1981 Biography 6.846052631578947 162441 1981 Comedy 5.845045045045046 956122 1981 Crime 5.961089494163424 474711 1981 Documentary 7.03962962962963 37504 1981 Drama 6.147728860936409 1603074 1981 Family 6.4010989010989015 44161 1981 Fantasy 6.125225225225226 289990 1981 History 6.777906976744186 169943 1981 Horror 5.396739130434782 1152995 1981 Music 6.748051948051948 43435 1981 Musical 6.2146341463414645 101656 1981 Mystery 5.957798165137614 309869 1981 Romance 6.090347490347491 314470 1981 Sci-Fi 5.3769230769230765 530667 1981 Short 6.46578947368421 25777 1981 Sport 6.13076923076923 155008 1981 Talk-Show 5.3 66 1981 Thriller 5.801324503311259 567666 1981 War 6.471014492753623 421177 1981 Western 5.844444444444444 5274 1982 Action 5.635260115606937 1766801 1982 Adult 5.581818181818182 17243 1982 Adventure 5.862566844919787 1563191 1982 Animation 6.621134020618556 171999 1982 Biography 6.745901639344263 298295 1982 Comedy 5.8629999999999995 1324314 1982 Crime 5.859493670886077 463886 1982 Documentary 7.127091633466136 94946 1982 Drama 6.097441077441078 2952986 1982 Family 6.3294999999999995 663459 1982 Fantasy 6.102362204724408 529586 1982 History 6.5694444444444455 295184 1982 Horror 5.285869565217391 1236649 1982 Music 6.8453608247422695 272920 1982 Musical 6.273267326732673 95800 1982 Mystery 5.863736263736264 759815 1982 Romance 6.049110320284696 430315 1982 Sci-Fi 5.601449275362318 2304931 1982 Short 6.620467836257311 23189 1982 Sport 5.833333333333333 239274 1982 Talk-Show 7.8 53 1982 Thriller 5.556603773584905 766075 1982 War 6.1245901639344265 34773 1982 Western 5.840000000000001 5060 1983 Action 5.590056818181818 1913215 1983 Adult 5.583928571428572 16180 1983 Adventure 5.686315789473683 1884514 1983 Animation 6.561860465116279 89817 1983 Biography 6.770422535211267 142036 1983 Comedy 5.901997503121098 1600255 1983 Crime 5.882499999999999 1585327 1983 Documentary 6.9978873239436625 41573 1983 Drama 6.122601416613007 2787479 1983 Family 6.430208333333334 233264 1983 Fantasy 6.043703703703704 1322446 1983 Game-Show 8.6 16 1983 History 6.6706666666666665 63581 1983 Horror 5.302721088435374 689190 1983 Music 6.52695652173913 166424 1983 Musical 6.21860465116279 175650 1983 Mystery 6.020792079207921 182659 1983 Romance 5.902027027027027 393047 1983 Sci-Fi 5.735714285714285 510984 1983 Short 6.449122807017545 21087 1983 Sport 6.092307692307693 31890 1983 Thriller 5.548148148148149 621404 1983 War 6.105952380952382 66575 1983 Western 5.677777777777778 4918 1984 Action 5.724705882352941 3466906 1984 Adult 5.565882352941176 14528 1984 Adventure 5.9567415730337085 2800510 1984 Animation 6.635483870967742 229675 1984 Biography 6.8674418604651155 544514 1984 Comedy 5.996992481203007 2644155 1984 Crime 5.901831501831502 1178363 1984 Documentary 7.0823754789272035 63056 1984 Drama 6.2390992167101835 2916267 1984 Family 6.4396475770925115 512903 1984 Fantasy 6.303999999999999 1068589 1984 History 6.8689189189189195 140702 1984 Horror 5.2875862068965525 968642 1984 Music 6.654400000000001 903220 1984 Musical 6.215789473684212 15367 1984 Mystery 5.764705882352941 172425 1984 Romance 5.967128027681661 666248 1984 Sci-Fi 5.610309278350516 1940796 1984 Short 6.5539772727272725 19606 1984 Sport 6.385185185185184 67311 1984 Thriller 5.740322580645161 515418 1984 War 5.9228571428571435 60473 1984 Western 6.238461538461538 7624 1985 Action 5.697727272727272 1594395 1985 Adult 5.829602888086643 19792 1985 Adventure 5.903664921465968 3211640 1985 Animation 6.486784140969164 127241 1985 Biography 6.656060606060605 186609 1985 Comedy 6.035029585798816 4310898 1985 Crime 5.844666666666665 710093 1985 Documentary 7.058174904942966 45052 1985 Drama 6.262004034969738 2970881 1985 Family 6.3838888888888885 524056 1985 Fantasy 6.0703125 435867 1985 History 6.89850746268

In [0]:
%sql
UPDATE gold.genre_analysis
SET avgRating = ROUND(avgRating, 3);

num_affected_rows 3073

Os códigos a seguir são feitas as consultas a fim de responder as perguntas originalmente elaboradas

In [0]:
%sql
SELECT DISTINCT *
FROM gold.genre_analysis
WHERE year = 1980
ORDER BY avgRating DESC;

year genre avgRating totalVotes 1980 Documentary 6.997 29097 1980 Biography 6.764 771122 1980 History 6.734 101770 1980 Music 6.651 87941 1980 Short 6.473 26113 1980 Animation 6.456 47625 1980 Family 6.399 114855 1980 War 6.29 137122 1980 Fantasy 6.193 1560092 1980 Romance 6.118 332014 1980 Drama 6.094 3002048 1980 Musical 6.062 63676 1980 Sport 6.016 537445 1980 Mystery 5.861 459015 1980 Adventure 5.851 2351648 1980 Comedy 5.811 1481880 1980 Crime 5.785 628009 1980 Action 5.758 1913657 1980 Western 5.679 44056 1980 Adult 5.596 17505 1980 Thriller 5.578 510763 1980 Sci-Fi 5.477 375351 1980 Horror 5.295 1913636

Aqui foi feito um recorte de um ano específico utilizando o comando _WHERE year = 1980_, para verificar as melhores avaliações por gênero. O _ORDER BY avgRating DESC_ ordena da maior para a menor nota. Observa-se notas mais elevadas pertencem ao gênero documentário. Horror é o último colocado.

In [0]:
%sql
SELECT DISTINCT *
FROM gold.genre_analysis
WHERE year = 1990
ORDER BY avgRating DESC;

year genre avgRating totalVotes 1990 Reality-TV 7.467 67 1990 Documentary 7.116 47469 1990 Music 7.048 83877 1990 History 6.703 75740 1990 Biography 6.694 1598286 1990 Animation 6.597 139980 1990 Musical 6.565 77287 1990 Short 6.434 24207 1990 Family 6.347 944198 1990 War 6.338 109235 1990 Adult 6.196 11098 1990 Drama 6.185 5788644 1990 Sport 6.067 270204 1990 Comedy 6.057 3770890 1990 Adventure 6.045 1936280 1990 Romance 5.938 1630067 1990 Fantasy 5.908 1146148 1990 Game-Show 5.9 18 1990 Mystery 5.892 362951 1990 Crime 5.813 3398405 1990 Western 5.793 345522 1990 Thriller 5.608 1782370 1990 Action 5.586 2749915 1990 Sci-Fi 5.37 1467440 1990 Horror 5.13 1305240

10 anos depois, no ano de 1990, documentário caiu uma posição e foi substituído por Reality TV, categoria que nem existia em 1990. No entanto, o número de votos é muito baixo. O fato do filtro ter selecionado apenas filmes na criação da tabela, indica um possível erro na categorização do Reality TV.
Observa-se que o valor da média de documentário aumentou, enquanto a média de horro decaiu, mantendo o gênero em último colocado.

In [0]:
%sql
SELECT *
FROM gold.genre_analysis
ORDER BY avgRating DESC
LIMIT 10;

year genre avgRating totalVotes 2015 Game-Show 9.9 10 1903 Sci-Fi 9.0 11 2003 News 8.65 66 1983 Game-Show 8.6 16 2022 News 8.6 2924 2005 Talk-Show 8.5 9 1960 Reality-TV 8.5 21 2005 Reality-TV 8.45 156 1993 News 8.4 31 1998 Talk-Show 8.4 142

Aqui são apresentadas as maiores notas. No entanto, é possível perceber um padrão de que há poucou votos em cada tupla, sendo uma métrica menos precisa pela pequena amostragem.

In [0]:
%sql
SELECT *
FROM gold.genre_analysis
WHERE totalVotes > 50000
ORDER BY avgRating DESC
LIMIT 10;

year genre avgRating totalVotes 2008 News 7.92 56033 2024 Short 7.695 64345 2023 Short 7.45 233516 1996 Music 7.442 219790 1961 Sport 7.413 90552 1995 Music 7.374 123715 2022 Short 7.374 208517 1989 Music 7.362 242762 2024 Documentary 7.354 509314 2022 Documentary 7.305 849212

Para obter uma amostragem maior, filtrou-se os resultados em que o número de votos foi superior à 50000

In [0]:
%sql
SELECT *
FROM gold.genre_analysis
WHERE genre = 'Drama'
ORDER BY avgRating DESC
LIMIT 50;

year genre avgRating totalVotes 2024 Drama 6.893 8043960 1888 Drama 6.8 6 2016 Drama 6.699 19822044 2023 Drama 6.696 12459619 2022 Drama 6.685 15410269 2012 Drama 6.674 19320243 2014 Drama 6.66 25873631 2017 Drama 6.658 19581096 2011 Drama 6.634 20291455 2015 Drama 6.617 19560153 2010 Drama 6.613 18686628 2013 Drama 6.599 20542861 2021 Drama 6.56 13227813 2018 Drama 6.549 16549687 2009 Drama 6.531 17093959 2020 Drama 6.495 9782747 1966 Drama 6.485 1167412 2019 Drama 6.482 20620483 1964 Drama 6.475 1227602 1962 Drama 6.474 2035121 1965 Drama 6.471 1469198 2006 Drama 6.469 20186058 1963 Drama 6.465 1270742 2008 Drama 6.462 20131865 2007 Drama 6.461 18066550 2005 Drama 6.457 14852913 1961 Drama 6.453 1371402 2004 Drama 6.451 16564184 1967 Drama 6.448 1399285 1959 Drama 6.387 1372757 1960 Drama 6.382 2134514 1956 Drama 6.373 1002102 1947 Drama 6.366 586545 1955 Drama 6.366 1100288 1948 Drama 6.365 1075476 1958 Drama 6.364 838773 1978 Drama 6.356 1396967 2003 Drama 6.354 12784037 2002 Drama 6.353 16351846 1957 Drama 6.342 2583193 1946 Drama 6.338 1360381 1954 Drama 6.331 1987088 1949 Drama 6.33 692663 1968 Drama 6.325 1476058 1952 Drama 6.291 727642 1951 Drama 6.29 971779 1992 Drama 6.283 4722043 1979 Drama 6.277 2646431 1931 Drama 6.276 581814 2001 Drama 6.272 14375162

Foi feita uma análise do gênero drama, avaliando quais os anos com as melhoras notas no gênero. Mais uma vez, o ano de 1888, com um número pequeno de notas, aparece com nota elevada, possivelmente por uma pequena amostragem.

In [0]:
%sql
SELECT *
FROM gold.genre_analysis
WHERE genre = 'Action'
ORDER BY avgRating DESC
LIMIT 50;

year genre avgRating totalVotes 2012 Action 6.361 15826469 1956 Action 6.335 48174 2024 Action 6.279 6812101 2014 Action 6.266 20360863 2016 Action 6.24 16698187 2011 Action 6.23 16529661 1943 Action 6.217 49185 2013 Action 6.215 16320595 1958 Action 6.202 77693 2015 Action 6.2 14161583 1948 Action 6.196 82800 2010 Action 6.169 15479130 2023 Action 6.168 9249618 1951 Action 6.167 58213 1957 Action 6.151 80802 1916 Action 6.134 5745 1947 Action 6.133 23768 1962 Action 6.125 348454 1945 Action 6.121 21111 1963 Action 6.106 292982 1964 Action 6.101 267845 2022 Action 6.094 11851088 1954 Action 6.093 447826 1923 Action 6.086 25880 1942 Action 6.075 55738 2017 Action 6.07 16144199 1918 Action 6.058 1650 1955 Action 6.056 53434 2020 Action 6.045 5527243 1927 Action 6.043 17431 2021 Action 6.042 12171234 1952 Action 6.035 61257 1960 Action 6.027 149823 2019 Action 6.025 12291687 1941 Action 6.013 43088 1940 Action 6.01 60742 1961 Action 6.009 234821 1914 Action 6.006 3051 1904 Action 6.0 4453 1969 Action 5.997 397701 1965 Action 5.994 270746 1949 Action 5.993 86651 1953 Action 5.981 85896 1944 Action 5.974 18991 2018 Action 5.971 14567904 1946 Action 5.967 11074 1950 Action 5.966 60944 1968 Action 5.963 303689 1925 Action 5.963 14728 1972 Action 5.961 465550

Quanto ao gênero ação, percebe-se uma maior avaliação nos filmes mais recentes.

In [0]:
%sql
SELECT year, SUM(totalVotes) AS total_votes_per_year
FROM gold.genre_analysis
GROUP BY year
ORDER BY total_votes_per_year DESC
LIMIT 50;

year total_votes_per_year 2014 130251052 2013 125331991 2016 109451559 2011 109298254 2012 108119573 2010 107872792 2008 105299012 2017 103790684 2015 103690345 2009 102858945 2019 101642170 2007 99098899 2018 97881074 2006 93602304 2004 90628781 2005 82199306 2002 77208028 2022 76809308 2001 75836536 2003 75501069 2021 74379436 1999 68058798 2023 66986496 2000 60634386 1997 54297633 1995 52433405 1998 52303758 2020 48381623 2024 45708695 1994 45534887 1993 36511883 1996 35851860 1990 29065538 1992 27404434 1991 26577112 1989 24190563 1988 23698479 1987 22899735 1984 20917298 1985 20441208 1986 20438717 1980 16506440 1982 16310744 1983 14543531 1979 12668459 1981 11662622 1975 11350596 1977 11130091 1972 10013595 1974 9783190

Por fim, o código acima mostra os anos com maior número de votos. Com o advento da internet e da popularização dos aparelhos eletrônicos, é esperado que filmes mais recentes apresentem maior número de votos devido ao momento que foram lançados. Não obstantante, o consumo de filmes é atemporal, especialmente na era do streaming, em que há acesso de forma fácil a diversos títulos. Assim, obras do século passado são consumidas nos dias atuais.

# AUTOAVALIAÇÃO
De maneira geral, avalio positivamente a minha experiência elaborando o trabalho. Incialmente, busquei base de dados referentes a premiações do mundo da música, como o Grammy. Pensei em fazer análises com gênero, mas nesse caso do álbum. No entanto, buscando no Kaggle tive dificuldades de encontrar uma base de dados consistida e tive um pouco de dificuldade com o CSV. Portando, resolvi ir para a do IMDb, que estava mais consolidada e que me permitiria fazer um trabalho parecido com o que originalmente pensei.

O primeiro problema que encontrei foi o com a coluna "isAdult". Pensei em corrigí-lo em vez de apagá-lo, mas como o dado não seria de meu interesse, optei por excluí-lo. Não sei se foi a ação mais prudente, considerando que futuramente poderia fazer o uso dele.
Outra questão que me ocorreu foi o fato da chave primária tconst ter perdido a sua unicidade quando explodi os gêneros. Julguei passível de ocorrer considerando que naõ afetaria os objetivos do projeto.

Tive certas dificuldades para elaborar alguns códigos, mas consultando consegui elaborá-los.

Ao final, fui percebendo possíveis falhas no modelo, especialmente considerando a amostragem dos dados. Filmes mais antigos apresentam poucas notas. Talvez fosse prudente limitar o período de análise ou excluir filmes com poucas notas. 

A criação do MPV foi muito interressante para colocar em prática o conteúdo aprendido e obter informações de maneira categórica de dados brutos, após o seu tratamente, foi muito enriquecedor. Consegui ter um gostinho do poder da análise de dados.